integrations/opennms-wsman-asset-provisioning-adapter/src/main/java/org/opennms/netmgt/provision/WsManAssetProvisioningAdapter.java

```diff
@@ -0,0 +1,463 @@
+                final WSManClient client = m_factory.getClient(endpoint);
+                LOG.debug("doAdd: m_config: " + m_config);
+
+                final OnmsAssetRecord asset = node.getAssetRecord();
+                m_config.getReadLock().lock();
+                try {
+                    for (final AssetField field : m_config.getAssetFieldsForAddress(ipaddress, vendor)) {
+                        try {
+                                final String value = fetchWsManAssetString(client, endpoint, field.getWqlObjs(), field.getFormatString());
+                                LOG.debug("doAdd: Setting asset field \" {} \" to value: {}", field.getName(), value);
+                                // Use Spring bean-accessor classes to set the field value
+                                final BeanWrapper wrapper = PropertyAccessorFactory.forBeanPropertyAccess(asset);
+                                try {
+                                        wrapper.setPropertyValue(field.getName(), value);
+                                } catch (final BeansException e) {
+                                        LOG.warn("doAdd: Could not set property \" {} \" on asset object {}", field.getName(), e.getMessage(), e);
+                                }
+                        } catch (final Throwable t) {
+                                // This exception is thrown if the WSMAN ENUM fails or an incorrect number of
+                                // parameters is returned by the agent or because of a misconfiguration.
+                                LOG.warn("doAdd: Could not set value for asset field \" {} \": {}", field.getName(), t.getMessage(), t);
+                        }
+                }
+                } finally {
+                    m_config.getReadLock().unlock();
+                }
+                                                                     
+        node.setAssetRecord(asset);
+        m_nodeDao.saveOrUpdate(node);
+                m_nodeDao.flush();
+        }
+
+        private static String fetchWsManAssetString(final WSManClient client, final WSManEndpoint endpoint, final List<WqlObj> wqlObjs, final String formatString) {
+
+                final List<String> aliases = new ArrayList<>();
+                final List<String> wqls = new ArrayList<>();
+                final List<String> resourceUris = new ArrayList<>();
+                final List<String> values = new ArrayList<>();
+                for (final WqlObj wqlobj : wqlObjs) {
+                	List<Node> nodes = Lists.newLinkedList();
+                        aliases.add(wqlobj.getAlias());
+                        wqls.add(wqlobj.getWql());
+                        resourceUris.add(wqlobj.getResourceUri());
+			client.enumerateAndPullUsingFilter(wqlobj.getResourceUri(), WSManConstants.XML_NS_WQL_DIALECT, wqlobj.getWql(), nodes, true);
+                 	values.add(nodes.get(0).getTextContent());
```

Comment:
What if `nodes` is empty?

Reply:
Is a conditional to check `nodes.size > 0` sufficient here, or is there else more appropriate?

Reply:
It's been a couple months, but to make it more explicit i could do:
```suggestion
                 	if (nodes.size() > 0) {
                                values.add(nodes.get(0).getTextContent());
                        } else {
                                values.add(null);
                        }
```
otherwise, `nodes.get(0)` should return `null`, and I'm checking for that when doing `values.get()` a few lines further down.

Reply:
That would work, but use `!nodes.isEmpty()` instead of `nodes.size() > 0`